File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxLength = 0
        charSet = set()
        left = 0

        for right, rightChar in enumerate(s):
            while rightChar in charSet:
                charSet.remove(s[left])
                left += 1
            currLength = right - left + 1
            maxLength = max(maxLength, currLength)
            charSet.add(rightChar)
        
        return maxLength

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let [left, right] = [0, 0];
    let maxLength = 0;
    const window = new Set();

    while (right < s.length) {
        const rightChar = s[right];

        if (window.has(rightChar)) {
            while (s[left] !== rightChar) {
                window.delete(s[left]);
                left += 1;
            }
            left += 1; 
        }

        window.add(rightChar);
        maxLength = Math.max(maxLength, window.size);
        right += 1;
    }

    return maxLength;
};
